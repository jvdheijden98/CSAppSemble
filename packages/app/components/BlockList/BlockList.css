.transitionGroup,
.transitionWrapper {
  display: flex;
  height: 100%;
  position: relative;
}

.transitionWrapper {
  left: 0;
  position: absolute;
  top: 0;
  width: 100%;
}

.pageEnter {
  opacity: 0;
}

.pageEnterActive {
  opacity: 1;
  transition: opacity 300ms ease-in;
}

.pageExit {
  opacity: 1;
}

.pageExitActive {
  opacity: 0;
  transition: opacity 300ms ease-in;
}
