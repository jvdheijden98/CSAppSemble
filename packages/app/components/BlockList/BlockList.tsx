import { Loader } from '@appsemble/react-components';
import { Action, Block as BlockType } from '@appsemble/types';
import React from 'react';
import { CSSTransition, TransitionGroup } from 'react-transition-group';

import { ShowDialogAction } from '../../types';
import Block from '../Block';
import styles from './BlockList.css';

export interface BlockListProps {
  actionCreators: Record<string, () => Action>;
  counter: number;
  currentPage?: number;
  blocks: BlockType[];
  data?: any;
  emitEvent(name: string, data: any): void;
  flowActions: {};
  offEvent(name: string, callback: Function): void;
  onEvent(name: string, callback: Function): void;
  showDialog: ShowDialogAction;
  transitions?: boolean;
}

interface BlockListState {
  blockStatus: Record<string, boolean>;
}

export default class BlockList extends React.Component<BlockListProps, BlockListState> {
  static defaultProps: Partial<BlockListProps> = {
    transitions: false,
    data: undefined,
  };

  state = {
    blockStatus: this.props.blocks.reduce<Record<string, boolean>>(
      (acc: Record<string, boolean>, block, index) => {
        acc[`${block.type}${index}`] = false;
        return acc;
      },
      {},
    ),
  };

  ready = (blockId: string): void => {
    this.setState(({ blockStatus }) => {
      return { blockStatus: { ...blockStatus, [blockId]: true } };
    });
  };

  render(): React.ReactNode {
    const {
      actionCreators,
      blocks,
      counter,
      currentPage,
      data,
      emitEvent,
      flowActions,
      offEvent,
      onEvent,
      showDialog,
      transitions,
    } = this.props;

    const { blockStatus } = this.state;
    const isLoading = Object.values(blockStatus).some(s => !s);

    const list = blocks.map((block, index) => {
      const content = (
        <Block
          // As long as blocks are in a static list, using the index as a key should be fine.
          // eslint-disable-next-line react/no-array-index-key
          key={`${index}.${counter}`}
          actionCreators={actionCreators}
          block={block}
          className={isLoading ? 'is-hidden' : ''}
          data={data}
          emitEvent={emitEvent}
          flowActions={flowActions}
          offEvent={offEvent}
          onEvent={onEvent}
          ready={() => this.ready(`${block.type}${index}`)}
          showDialog={showDialog}
        />
      );

      return transitions ? (
        <CSSTransition
          // Since blocks are in a static list, using the index as a key should be fine.
          // eslint-disable-next-line react/no-array-index-key
          key={`${currentPage}.${index}.${counter}`}
          classNames={{
            enter: styles.pageEnter,
            enterActive: styles.pageEnterActive,
            exit: styles.pageExit,
            exitActive: styles.pageExitActive,
          }}
          timeout={300}
        >
          <div className={styles.transitionWrapper}>{content}</div>
        </CSSTransition>
      ) : (
        content
      );
    });

    return (
      <>
        {isLoading && <Loader />}
        {transitions ? (
          <TransitionGroup className={styles.transitionGroup}>{list}</TransitionGroup>
        ) : (
          list
        )}
      </>
    );
  }
}
