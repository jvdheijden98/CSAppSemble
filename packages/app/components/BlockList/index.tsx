export { default, BlockListProps } from './BlockList';
