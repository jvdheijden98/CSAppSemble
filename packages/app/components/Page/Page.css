.error {
  margin: 20vh 8px 8px;
}
