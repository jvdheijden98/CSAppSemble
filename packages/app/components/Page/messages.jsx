import { defineMessages } from 'react-intl';

export default defineMessages({
  error: 'There was a problem loading this page. Weâ€™re sorry.',
});
