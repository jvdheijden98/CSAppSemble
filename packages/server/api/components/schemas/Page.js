export default {
  oneOf: [
    {
      type: 'object',
      description: 'This describes what a page will look like in the app.',
      required: ['name', 'blocks'],
      properties: {
        name: {
          type: 'string',
          maxLength: 50,
          description: `The name of an app.

          This will be displayed on the top of the page and in the side menu.
        `,
        },
        type: {
          type: 'string',
          enum: ['flow', 'page', 'tabs'],
        },
        icon: {
          type: 'string',
          description: `An optional icon from the fontawesome icon set

            This will be displayed in the navigation menu.
          `,
        },
        parameters: {
          type: 'array',
          description:
            'Page parameters can be used for linking to a page that should display a single resource.',
          items: {
            type: 'string',
            minLength: 1,
            maxLength: 30,
          },
        },
        scope: {
          type: 'array',
          description: `Specify which authentication scopes are needed to view this page.

          If a user either isn’t logged in, or doesn’t have sufficient permissions to view the page,
          they will be prompted with a login screen.
        `,
          items: {
            enum: ['*'],
          },
        },
        theme: {
          $ref: '#/components/schemas/Theme',
        },
        blocks: {
          type: 'array',
          minItems: 1,
          description: 'The blocks that make up a page.',
          items: {
            $ref: '#/components/schemas/Block',
          },
        },
        subPages: {
          type: 'array',
          minItems: 1,
          description: "Sub pages belonging to this page's flow.",
          items: {
            $ref: '#/components/schemas/SubPage',
          },
        },
        actions: {
          type: 'object',
          description: 'A mapping of actions that can be fired by the page to action handlers.',
          additionalProperties: {
            type: 'object',
          },
        },
        navigation: {
          $ref: '#/components/schemas/App/properties/definition/properties/navigation',
        },
        hideFromMenu: {
          type: 'boolean',
          description: 'Whether or not the page should be displayed in navigational menus.',
          default: false,
        },
      },
    },
    {
      type: 'object',
      description: 'This describes what a page will look like in the app.',
      required: ['name', 'subPages'],
      properties: {
        name: {
          type: 'string',
          maxLength: 50,
          description: `The name of an app.

          This will be displayed on the top of the page and in the side menu.
        `,
        },
        type: {
          type: 'string',
          enum: ['flow', 'page', 'tabs'],
        },
        icon: {
          type: 'string',
          description: `An optional icon from the fontawesome icon set

            This will be displayed in the navigation menu.
          `,
        },
        parameters: {
          type: 'array',
          description:
            'Page parameters can be used for linking to a page that should display a single resource.',
          items: {
            type: 'string',
            minLength: 1,
            maxLength: 30,
          },
        },
        scope: {
          type: 'array',
          description: `Specify which authentication scopes are needed to view this page.

          If a user either isn’t logged in, or doesn’t have sufficient permissions to view the page,
          they will be prompted with a login screen.
        `,
          items: {
            enum: ['*'],
          },
        },
        theme: {
          $ref: '#/components/schemas/Theme',
        },
        subPages: {
          type: 'array',
          minItems: 1,
          description: "Sub pages belonging to this page's flow.",
          items: {
            $ref: '#/components/schemas/SubPage',
          },
        },
        actions: {
          type: 'object',
          description: 'A mapping of actions that can be fired by the page to action handlers.',
          additionalProperties: {
            type: 'object',
          },
        },
        navigation: {
          $ref: '#/components/schemas/App/properties/definition/properties/navigation',
        },
        hideFromMenu: {
          type: 'boolean',
          description: 'Whether or not the page should be displayed in navigational menus.',
          default: false,
        },
      },
    },
  ],
};
