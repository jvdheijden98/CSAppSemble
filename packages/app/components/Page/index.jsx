import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import { getBlockDefs } from '../../actions/blockDefs';
import Page from './Page';

function mapStateToProps(state, ownProps) {
  return {
    definition: state.app.definition,
    user: state.user.user,
    pending: !!state.blockDefs.pending.length,
    hasErrors:
      ownProps.page.type && (ownProps.page.type === 'flow' || ownProps.page.type === 'tabs')
        ? ownProps.page.subPages.some(sub =>
            sub.blocks.some(block => state.blockDefs.errored.has(block.type)),
          )
        : ownProps.page.blocks.some(block => state.blockDefs.errored.has(block.type)),
  };
}

export default withRouter(
  connect(mapStateToProps, {
    getBlockDefs,
  })(Page),
);
