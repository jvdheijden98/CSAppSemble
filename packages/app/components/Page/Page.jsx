import { Loader } from '@appsemble/react-components';
import EventEmitter from 'events';
import PropTypes from 'prop-types';
import React from 'react';
import { FormattedMessage } from 'react-intl';

import checkScope from '../../utils/checkScope';
import makeActions from '../../utils/makeActions';
import BlockList from '../BlockList';
import FlowPage from '../FlowPage';
import Login from '../Login';
import PageDialog from '../PageDialog';
import TabsPage from '../TabsPage';
import TitleBar from '../TitleBar';
import messages from './messages';
import styles from './Page.css';

/**
 * Render an app page definition.
 */
export default class Page extends React.Component {
  flowActions = {
    next: async data => {
      const { currentPage } = this.state;
      const { page } = this.props;
      const { subPages } = page;

      if (currentPage + 1 === subPages.length) {
        await this.actions.onFlowFinish.dispatch(data);
        return data;
      }

      this.setState({ data, currentPage: currentPage + 1 });
      return data;
    },

    finish: async data => {
      await this.actions.onFlowFinish.dispatch(data);
      this.setState({ data });
      return data;
    },

    back: async data => {
      const { currentPage } = this.state;

      if (currentPage <= 0) {
        // Don't do anything if a previous page does not exist
        return data;
      }

      this.setState({ data, currentPage: currentPage - 1 });
      return data;
    },

    cancel: async data => {
      await this.actions.onFlowCancel.dispatch(data);
      this.setState({ data });
    },
  };

  static propTypes = {
    appId: PropTypes.number.isRequired,
    definition: PropTypes.shape().isRequired,
    getBlockDefs: PropTypes.func.isRequired,
    hasErrors: PropTypes.bool.isRequired,
    history: PropTypes.shape().isRequired,
    match: PropTypes.shape().isRequired,
    /**
     * The page definition to render
     */
    page: PropTypes.shape().isRequired,
    pending: PropTypes.bool.isRequired,
    user: PropTypes.shape(),
  };

  static defaultProps = {
    user: null,
  };

  state = {
    dialog: null,
    actions: {},
    counter: 0,
    currentPage: 0,
    data: {},
  };

  constructor(props) {
    super(props);
    this.setupEvents();
  }

  componentDidMount() {
    const { appId, definition, getBlockDefs, page, history } = this.props;

    this.applyBulmaThemes(definition, page);
    this.setupEvents();

    if (page.type === 'flow') {
      const actions = makeActions(
        appId,
        { actions: { onFlowFinish: {}, onFlowCancel: {} } },
        definition,
        page,
        history,
        this.showDialog,
        {
          emit: this.emitEvent,
          off: this.offEvent,
          on: this.onEvent,
        },
        {},
        this.flowActions,
      );

      this.actions = actions;
    }

    const blocks = [
      ...(page.type === 'tabs' || page.type === 'flow'
        ? page.subPages.map(f => f.blocks).flat()
        : []),
      ...(!page.type || page.type === 'page' ? page.blocks : []),
    ];

    const actionBlocks = blocks
      .filter(block => block.actions)
      .map(block => {
        return Object.entries(block.actions)
          .filter(([, action]) => action.type === 'dialog')
          .map(([, action]) => action.blocks);
      })
      .flat(2);

    getBlockDefs([...new Set([...blocks, ...actionBlocks])]);
  }

  static getDerivedStateFromProps(props, state) {
    if (state.prevPage !== props.page) {
      return { ...state, prevPage: props.page, counter: state.counter + 1 };
    }

    // Nothing to update.
    return null;
  }

  componentDidUpdate({ page: prevPage }, { prevCurrentPage }) {
    const { definition, getBlockDefs, page } = this.props;
    const { currentPage } = this.state;

    if (page !== prevPage || prevCurrentPage !== currentPage) {
      this.teardownEvents();
      this.setupEvents();

      if (page.type === 'flow' || page.type === 'tabs') {
        getBlockDefs(
          [
            ...new Set(
              page.subPages
                .map(f => f.blocks)
                .flat()
                .map(b => JSON.stringify({ type: b.type, version: b.version })),
            ),
          ].map(block => JSON.parse(block)),
        );
      } else {
        getBlockDefs(page.blocks);
      }

      this.applyBulmaThemes(definition, page);
    }
  }

  componentWillUnmount() {
    this.teardownEvents();
  }

  setupEvents() {
    const ee = new EventEmitter();
    this.emitEvent = (name, data) => ee.emit(name, data);
    this.offEvent = (name, callback) => ee.off(name, callback.bind());
    this.onEvent = (name, callback) => ee.on(name, callback.bind());
    this.ee = ee;
  }

  createBulmaQueryString = () => {
    const { definition, page } = this.props;
    const params = { ...definition.theme, ...page.theme };
    const queryStringParams = new URLSearchParams(params);
    queryStringParams.sort();

    return queryStringParams.toString();
  };

  applyBulmaThemes = (definition, page) => {
    const bulmaStyle = document.getElementById('bulma-style-app');
    const [bulmaUrl] = bulmaStyle.href.split('?');
    bulmaStyle.href =
      definition.theme || page.theme ? `${bulmaUrl}?${this.createBulmaQueryString()}` : bulmaUrl;
  };

  showDialog = dialog => {
    this.setState({ dialog });
    return () => {
      this.setState({ dialog: null });
    };
  };

  teardownEvents() {
    if (this.ee) {
      this.ee.removeAllListeners();
      this.ee = null;
      this.emitEvent = null;
      this.offEvent = null;
      this.onEvent = null;
    }
  }

  render() {
    const { hasErrors, page, user, pending } = this.props;
    const { dialog, counter, currentPage, data } = this.state;
    const { type } = page;

    if (!checkScope(page.scope, user)) {
      return (
        <>
          <TitleBar>{page.name}</TitleBar>
          <Login />
        </>
      );
    }

    if (hasErrors) {
      return (
        <p className={styles.error}>
          <FormattedMessage {...messages.error} />
        </p>
      );
    }

    if (pending) {
      return <Loader />;
    }

    let component;
    switch (type) {
      case 'flow':
        component = (
          <FlowPage
            blocks={page.subPages[currentPage].blocks}
            counter={counter}
            currentPage={currentPage}
            data={data}
            emitEvent={this.emitEvent}
            flowActions={this.flowActions}
            offEvent={this.offEvent}
            onEvent={this.onEvent}
            showDialog={this.showDialog}
            subPages={page.subPages}
          />
        );
        break;
      case 'tabs':
        component = (
          <TabsPage
            counter={counter}
            data={data}
            emitEvent={this.emitEvent}
            flowActions={this.flowActions}
            offEvent={this.offEvent}
            onEvent={this.onEvent}
            showDialog={this.showDialog}
            subPages={page.subPages}
          />
        );
        break;
      default:
        component = (
          <BlockList
            blocks={page.blocks}
            counter={counter}
            data={data}
            emitEvent={this.emitEvent}
            flowActions={this.flowActions}
            offEvent={this.offEvent}
            onEvent={this.onEvent}
            showDialog={this.showDialog}
          />
        );
    }

    return (
      <>
        <TitleBar>{page.name}</TitleBar>
        {component}
        <PageDialog
          dialog={dialog}
          emitEvent={this.emitEvent}
          offEvent={this.offEvent}
          onEvent={this.onEvent}
        />
      </>
    );
  }
}
