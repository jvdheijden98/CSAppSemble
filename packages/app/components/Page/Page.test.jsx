import { shallow } from 'enzyme';
import React from 'react';

import BlockList from '../BlockList';
import Page from './Page';

beforeEach(() => {
  jest.spyOn(document, 'getElementById').mockReturnValue(document.createElement('link'));
});

it('should render the blocks for a page', () => {
  const page = { name: 'Test Page', blocks: [{}, {}] };
  const definition = { pages: [page] };
  const wrapper = shallow(
    <Page
      appId={1}
      definition={definition}
      getBlockDefs={() => {}}
      hasErrors={false}
      history={{}}
      match={{}}
      page={page}
    />,
  );

  expect(wrapper).toMatchSnapshot();
});

it('should render tabs pages', () => {
  const page = {
    name: 'Test Page',
    type: 'tabs',
    subPages: [
      { name: 'Sub A', blocks: [{}, {}] },
      { name: 'Sub B', blocks: [{}, {}] },
    ],
  };
  const definition = { pages: [page] };
  const wrapper = shallow(
    <Page
      appId={1}
      definition={definition}
      getBlockDefs={() => {}}
      hasErrors={false}
      history={{}}
      match={{}}
      page={page}
    />,
  );

  expect(wrapper).toMatchSnapshot();
});

it('should render flow page', () => {
  const page = {
    name: 'Test Page',
    type: 'flow',
    subPages: [
      { name: 'Sub A', blocks: [{}, {}] },
      { name: 'Sub B', blocks: [{}, {}] },
    ],
  };
  const definition = { pages: [page] };
  const wrapper = shallow(
    <Page
      appId={1}
      definition={definition}
      getBlockDefs={() => {}}
      hasErrors={false}
      history={{}}
      match={{}}
      page={page}
    />,
  );

  expect(wrapper).toMatchSnapshot();
});

it('should call event listeners when an event is emitted', () => {
  const page = { name: 'Test Page', blocks: [{}] };
  const definition = { pages: [page] };
  const wrapper = shallow(
    <Page
      appId={1}
      definition={definition}
      getBlockDefs={() => {}}
      hasErrors={false}
      history={{}}
      match={{}}
      page={page}
    />,
  );

  const blockProps = wrapper.find(BlockList).props();
  const spy = jest.fn();
  blockProps.onEvent('test', spy);
  blockProps.emitEvent('test', 'data');
  expect(spy).toHaveBeenCalledWith('data');
});
