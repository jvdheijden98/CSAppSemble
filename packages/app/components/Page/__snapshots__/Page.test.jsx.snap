// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render flow page 1`] = `
<Fragment>
  <TitleBar>
    Test Page
  </TitleBar>
  <FlowPage
    blocks={
      Array [
        Object {},
        Object {},
      ]
    }
    counter={1}
    currentPage={0}
    data={Object {}}
    emitEvent={[Function]}
    flowActions={
      Object {
        "back": [Function],
        "cancel": [Function],
        "finish": [Function],
        "next": [Function],
      }
    }
    offEvent={[Function]}
    onEvent={[Function]}
    showDialog={[Function]}
    subPages={
      Array [
        Object {
          "blocks": Array [
            Object {},
            Object {},
          ],
          "name": "Sub A",
        },
        Object {
          "blocks": Array [
            Object {},
            Object {},
          ],
          "name": "Sub B",
        },
      ]
    }
  />
  <Connect(PageDialog)
    dialog={null}
    emitEvent={[Function]}
    offEvent={[Function]}
    onEvent={[Function]}
  />
</Fragment>
`;

exports[`should render tabs pages 1`] = `
<Fragment>
  <TitleBar>
    Test Page
  </TitleBar>
  <withRouter(TabsPage)
    counter={1}
    data={Object {}}
    emitEvent={[Function]}
    flowActions={
      Object {
        "back": [Function],
        "cancel": [Function],
        "finish": [Function],
        "next": [Function],
      }
    }
    offEvent={[Function]}
    onEvent={[Function]}
    showDialog={[Function]}
    subPages={
      Array [
        Object {
          "blocks": Array [
            Object {},
            Object {},
          ],
          "name": "Sub A",
        },
        Object {
          "blocks": Array [
            Object {},
            Object {},
          ],
          "name": "Sub B",
        },
      ]
    }
  />
  <Connect(PageDialog)
    dialog={null}
    emitEvent={[Function]}
    offEvent={[Function]}
    onEvent={[Function]}
  />
</Fragment>
`;

exports[`should render the blocks for a page 1`] = `
<Fragment>
  <TitleBar>
    Test Page
  </TitleBar>
  <BlockList
    blocks={
      Array [
        Object {},
        Object {},
      ]
    }
    counter={1}
    data={Object {}}
    emitEvent={[Function]}
    flowActions={
      Object {
        "back": [Function],
        "cancel": [Function],
        "finish": [Function],
        "next": [Function],
      }
    }
    offEvent={[Function]}
    onEvent={[Function]}
    showDialog={[Function]}
    transitions={false}
  />
  <Connect(PageDialog)
    dialog={null}
    emitEvent={[Function]}
    offEvent={[Function]}
    onEvent={[Function]}
  />
</Fragment>
`;
